### PR TITLE
Improve equivalence checking of pointer arithmetic expressions

### PR DESCRIPTION
**Pull request by ppashakhanloo**
A new equivalence checking function has been added in this change
(EqualPointerArithOffset). This function is used for comparing two
bounds (i.e., Base + Offset) where one bound is following pointer
arithmetic form and the other one is in bytewise arithmetic form.
The assumption is that overflow checking is performed inside the *alloc function.

This pull request also contains new tests.

### CODE REVIEW

File: test/CheckedC/static-checking/bounds-decl-checking-cs.c [hunk 1]
```diff
@@ -2,9 +2,8 @@
...
 // true or false.
 //
-// RUN: %clang -cc1 -fcheckedc-extension -verify %s
+// RUN: %clang_cc1 -fcheckedc-extension -verify %s
 
 #pragma CHECKED_SCOPE ON
 
-#include "bounds-decl-checking.c"
-
+#include "bounds-decl-checking.c"
```
Comment (mgrang) [inline test/CheckedC/static-checking/bounds-decl-checking-cs.c]: Please add back the newline here.

File: test/CheckedC/static-checking/bounds-decl-checking-cs.c [hunk 2]
```diff
@@ -2,9 +2,8 @@
 // produces warnings about bounds declarations that are not provably
 // true or false.
 //
-// RUN: %clang -cc1 -fcheckedc-extension -verify %s
+// RUN: %clang_cc1 -fcheckedc-extension -verify %s
```
Comment (mgrang) [inline test/CheckedC/static-checking/bounds-decl-checking-cs.c]: Is there a reason to change `%clang -cc1` to `%clang_cc1`? AFAIK both of these are equivalent, no?

Comment (ppashakhanloo) [reply test/CheckedC/static-checking/bounds-decl-checking-cs.c]: Apparently, they are not the same. I needed to include <stddef.h>. It does not seem to work with `clang -cc1`. It says it cannot find stddef.h.
Please let me know if there is something better I could do here.

File: test/CheckedC/static-checking/bounds-decl-checking-cs.c (hunk 1 above)
Comment (ppashakhanloo) [reply test/CheckedC/static-checking/bounds-decl-checking-cs.c]: Done.

File: test/CheckedC/static-checking/bounds-decl-checking-cs.c (hunk 2 above)
Comment (mgrang) [reply test/CheckedC/static-checking/bounds-decl-checking-cs.c]: `clang` is the driver while `clang_cc1` is the clang frontend. Basically, clang driver will take in all the flags/options and will expand them into a clang -cc1 command line. So if you want to use `clang -cc1` then you would need to provide the path to the includes. So I guess what you did here is correct in that we do not need to go through the `clang` indirection and can directly invoke `clang_cc1`.

File: lib/Sema/SemaBounds.cpp [hunk 3]
```diff
@@ -2058,6 +2060,126 @@ namespace {
       return BaseRange::Kind::ConstantSized;
     }
 
+    // In this function, the goal is to compare two expressions: `Base1 + Offset1` and `Base2 + Offset2`
```
Comment (dtarditi) [inline lib/Sema/SemaBounds.cpp]: I think a more general approach would be to try to put each arithmetic expression into a standard form involving integer arithmetic.  If that is possible, you can then compare the standard forms.   The standard form would be base + (const op e), where op is either signed multiplication or unsigned multiplication, and base, const, and e are all interpreted as integers.   You could have variables that represent each of the components of the standard form.

Given `array_ptr<int>` p:
- For `((char *) p) + e1`, the standard form is `base = ((char *) p, const = 1, e = e1.`  op depends on the signedness of e1.
- For `p + e2`. the standard form is `base = p, const = sizeof(int), e = e2`.  op depends on the signedness of `e2`.
- For `((unsigned char *) p) + e1`, the standard from is `base = (unsigned char *) p, const = 1, e = e1`.  op depends on the signedness of `e1`.

To check that two expressions to produce identical results, we would check that the multiplication operators have the same signedness.   At this time, I don't think we should try to handle the case where one multiplication operator is signed and the other is unsigned (we would need additional facts to prove that the two operations produce the same value: that the signed integer is >= 0 and that the multiplication does not overflow).

**Comment by ppashakhanloo** [comment]
@dtarditi I made some changes to the code. It is ready for review.
There is now a new function CreateStandardForm(...) which basically expands the pointer arithmetic to bytewise arithmetic. I added some new tests and updated some comments as well.

File: lib/Sema/SemaBounds.cpp [hunk 4]
```diff
@@ -2058,6 +2058,97 @@ namespace {
...
+    // 1. For (char *)p + e1 or (unsigned char *)p + e1, ConstantPart = 1, VariablePart = e1.
+    // 2. For p + e1, ConstantPart = sizeof(T), VariablePart = e1.
+    //
+    // Note that another way to interpret the functionality of this function is that it expands
+    // pointer arithmetic to bytewise arithmetic.
+    static bool CreateStandardForm(ASTContext &Ctx, Expr *Base, Expr *Offset, llvm::APSInt &ConstantPart, bool &IsOpSigned, Expr *&VariablePart) {
+      ConstantPart = llvm::APSInt(Ctx.getTargetInfo().getPointerWidth(0), false);
+
+      if (Base->getType()->getPointeeOrArrayElementType()->getCanonicalTypeInternal()->isCharType()) {
+        if (BinaryOperator *BO = dyn_cast<BinaryOperator>(Offset->IgnoreParenCasts())) {
```
Comment (dtarditi) [inline lib/Sema/SemaBounds.cpp]: I think you want to call IgnoreParens() here.   The code should not ignore other casts.

File: lib/Sema/SemaBounds.cpp [hunk 5]
```diff
@@ -2058,6 +2058,97 @@ namespace {
...
+          if (BO->getRHS()->IgnoreParenCasts()->isIntegerConstantExpr(ConstantPart, Ctx))
+            VariablePart = BO->getLHS()->IgnoreParenCasts();
```
Comment (dtarditi) [inline lib/Sema/SemaBounds.cpp]: I think you can delete the call to IgnoreParenCasts().   The code should not ignore casts, and it does not need to ignore parenthesis.

File: lib/Sema/SemaBounds.cpp [hunk 6]
```diff
@@ -2058,6 +2058,97 @@ namespace {
...
+    // 2. For p + e1, ConstantPart = sizeof(T), VariablePart = e1.
+    //
+    // Note that another way to interpret the functionality of this function is that it expands
+    // pointer arithmetic to bytewise arithmetic.
+    static bool CreateStandardForm(ASTContext &Ctx, Expr *Base, Expr *Offset, llvm::APSInt &ConstantPart, bool &IsOpSigned, Expr *&VariablePart) {
+      ConstantPart = llvm::APSInt(Ctx.getTargetInfo().getPointerWidth(0), false);
+
+      if (Base->getType()->getPointeeOrArrayElementType()->getCanonicalTypeInternal()->isCharType()) {
+        if (BinaryOperator *BO = dyn_cast<BinaryOperator>(Offset->IgnoreParenCasts())) {
+          if (BO->getRHS()->IgnoreParenCasts()->isIntegerConstantExpr(ConstantPart, Ctx))
```
Comment (dtarditi) [inline lib/Sema/SemaBounds.cpp]: You can remove the IgnoreParenCasts() here.

Comment (ppashakhanloo) [reply lib/Sema/SemaBounds.cpp]: Done

File: lib/Sema/SemaBounds.cpp (hunk 5 above)
Comment (ppashakhanloo) [reply lib/Sema/SemaBounds.cpp]: Done

File: lib/Sema/SemaBounds.cpp (hunk 4 above)
Comment (ppashakhanloo) [reply lib/Sema/SemaBounds.cpp]: Done

File: lib/Sema/SemaBounds.cpp [hunk 7]
```diff
@@ -2058,6 +2058,97 @@ namespace {
...
+    // Given array_ptr<T> p:
+    // 1. For (char *)p + e1 or (unsigned char *)p + e1, ConstantPart = 1, VariablePart = e1.
+    // 2. For p + e1, ConstantPart = sizeof(T), VariablePart = e1.
+    //
+    // Note that another way to interpret the functionality of this function is that it expands
+    // pointer arithmetic to bytewise arithmetic.
+    static bool CreateStandardForm(ASTContext &Ctx, Expr *Base, Expr *Offset, llvm::APSInt &ConstantPart, bool &IsOpSigned, Expr *&VariablePart) {
+      ConstantPart = llvm::APSInt(Ctx.getTargetInfo().getPointerWidth(0), false);
+
+      if (Base->getType()->getPointeeOrArrayElementType()->getCanonicalTypeInternal()->isCharType()) {
```
[the remark this answers is not in the document]
Comment (ppashakhanloo) [reply lib/Sema/SemaBounds.cpp]: Done

File: lib/Sema/SemaBounds.cpp [hunk 8]
```diff
@@ -2058,6 +2058,97 @@ namespace {
...
+          ConstantPart = llvm::APSInt(ConstantPart, false);
+          if (ConstantPart.getBitWidth() < Ctx.getTargetInfo().getPointerWidth(0))
+            ConstantPart = ConstantPart.extend(Ctx.getTargetInfo().getPointerWidth(0));
+        } else
+          goto exit;
+        return true;
+
+      exit:
+        VariablePart = Offset->IgnoreParenCasts();
+        ConstantPart = 1;
```
[the remark this answers is not in the document]
Comment (ppashakhanloo) [reply lib/Sema/SemaBounds.cpp]: Great, I was very unhappy with the way that I had found! I replaced it with your suggestion.

**Comment by ppashakhanloo** [comment]
@dtarditi I have addressed all the above comments. The code is ready for review.
There are some more updates that can be found in checkedc repo. I opened a pull request at [microsoft/checkedc](https://github.com/microsoft/checkedc/pull/376).

**Pull request merged**